src/librustdoc/html/format.rs

```diff
@@ -1089,19 +1089,32 @@ impl Function<'_> {
 
 impl clean::Visibility {
     crate fn print_with_space(&self) -> impl fmt::Display + '_ {
+        use rustc_span::symbol::kw;
+
         display_fn(move |f| match *self {
             clean::Public => f.write_str("pub "),
             clean::Inherited => Ok(()),
             clean::Visibility::Crate => write!(f, "pub(crate) "),
+            // If this is `pub(crate)`, `path` will be empty.
+            clean::Visibility::Restricted(did, _) if did.index == CRATE_DEF_INDEX => {
```

Comment:
Looks like the logic here and in "`Clean for Visibility`" is duplicated.
Normalization "pub (in path)" -> ""/"path(crate)", and "pub(crate)" -> "" should either be done entirely here (perferable?) or entirely there.

Reply:
> Normalization "pub (in path)" -> ""/"path(crate)", and "pub(crate)" -> "" should either be done entirely here (perferable?)

It would be nice to do it here, but there's not enough context left: `print_with_space` has neither a `tcx` nor the `HirId` of the original item, so it can't call `tcx.parent_module(hir_id)` to know whether to display the visibility as `pub(self)` or not. I'll add this in `Clean for ty::Visibility` and `Clean for hir::Visibility` instead, moving the previous `"self"/"super"` logic to `hir::Visibility`.

Reply:
Actually, this information isn't available there either :/ I have just the visibility and not the `Item` it's being calculated for. I'll try turning both of those into a free function, maybe?